### PR TITLE
[Feature][New Model]Add Kimi-Audio model integration

### LINKED ISSUE #1824: [New Model]: Kimi-Audio-7B
### The model to consider.

https://github.com/MoonshotAI/Kimi-Audio

### The closest model vllm-omni already supports.

Qwen3-TTS

### What's your difficulty of supporting the model you want?

_No response_

### Use case and motivation

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by HelloWorldU**
## Purpose

Add Kimi-Audio integration with input preparation, a dual-stream AR stage, and a separate acoustic decoding stage.

- Input: reuse framework tokenization and multimodal processing, with Kimi-specific message rendering and GLM/Whisper input encoding.
- AR: integrate the delayed text/audio streams, per-request sampling and native weight loading. The unified entry constructs stages through the model registry; AR pipeline-parallel support is implemented but not GPU-validated.
- Output: integrate Kimi's flow-matching decoder and streaming state, reuse Omni's BigVGAN implementation, and connect offline generation, chat/speech serving and cross-stage `async_chunk` streaming.

Model implementation lives under `model_executor/models/kimi_audio/`; pipeline and deployment configuration, stage input processing, renderer and speech adapter use the corresponding framework extension points. A parameterized `examples/offline_inference/kimi_audio/end2end.py` demonstrates text and audio requests and saving their complete outputs. Reference tests were reduced to component boundaries and a shared-server online suite.

Closes https://github.com/vllm-project/vllm-omni/issues/1824.

### Remaining work

All done.

## Test Plan

The following integration tests are planned as pipeline and serving support become available:

1. **Single request:** cover supported text/audio input-output modes and multi-turn inputs. Verify text termination, waveform shape, non-empty audio where expected, and the 24 kHz sample rate.
2. **Batched and concurrent requests:** use requests with different lengths and input modes. Verify independent completion, no cross-request state leakage, and correct cleanup after completion or cancellation.
3. **`async_chunk` streaming:** verify that audio arrives before AR generation finishes, chunks are delivered in order without replay or omission, and the final partial chunk is flushed. Check playback continuity at chunk boundaries.
4. **Offline and online E2E:** exercise the registered pipeline and serving interfaces with pretrained weights on GPU. Reuse existing test infrastructure, including the shared `omni_server` fixture and local audio fixtures, and follow the project's CI marker conventions.

Retain focused component-level reference tests for input encoding, dual-stream forward and sampling, stage handoff, and acoustic decoding.

**vLLM Version:** 0.28.0; PyTorch 2.13.0+cu132; Python 3.12.14.

## Test Result

All planned single-GPU integration tests completed successfully on a single NVIDIA A800, including offline and online generation, streaming, concurrent requests, and cancellation/recovery.

One test case showed repetitive generation and reached the 512-step limit. A comparison run with the official Kimi-Audio reference implementation reproduced the same behavior under the same prompt and sampling settings, indicating that this is model-generation behavior rather than a vLLM-Omni-specific issue.

Multi-GPU, performance, stress, and fault-injection testing were not covered.

Assisted by: GPT-6 Astra.

### CODE REVIEW

**Comment by HelloWorldU** [comment]
Hi there! The PR is ready for review. Two points may deserve particular attention:
I. The design of `ids.output` in `vllm_omni/outputs/output_processor.py`
II. `include_stop_str_in_output=True` in  `vllm_omni/outputs/output_processor.py`
Both come from the same architectural constraint in the Kimi-Audio integration.

**Comment by HelloWorldU** [comment]
Hi @linyueqian, I’ve fixed the PR-related CI failure. The remaining failures were a Qwen-Image issue on AMD (possibly related to #7555) and insufficient disk space on Intel. CI hasn’t restarted after the push—could you please trigger it for the latest commit?

File: vllm_omni/model_executor/models/kimi_audio/sampling.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        if penalty > 1.0 and len(history) > window:
+            logits = logits.clone()
+            recent = torch.as_tensor(history[-window:], device=logits.device, dtype=torch.long)
+            scores = logits.gather(0, recent)
+            scores = torch.where(scores < 0, scores * penalty, scores / penalty)
+            logits.scatter_(0, recent, scores)
+
+        logprobs = torch.log_softmax(logits.unsqueeze(0), dim=-1, dtype=torch.float)
+        if temperature > 1e-6:
+            logprobs = logprobs / temperature
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/kimi_audio/sampling.py]: Apply temperature before log_softmax; low temperatures currently underflow every sampling probability.

File: vllm_omni/model_executor/models/kimi_audio/kimi_audio_ar_stage.py [hunk 2]
```diff
@@ -0,0 +1,876 @@
...
+                    raise ValueError("Kimi-Audio audio spans require native multimodal embeddings")
+                audio = self.embed_tokens(input_ids[: stop - offset])
+            else:
+                audio = input_embeds[: stop - offset]
+            text = torch.tensor(payload["text_token_ids"][offset:stop], device=input_ids.device, dtype=torch.long)
+            ids.extend(prompt_ids)
+            embeds.append(audio + self.embed_tokens(text))
+        if end > prompt_len:
+            start, stop = max(0, offset - prompt_len), end - prompt_len
+            scheduled = input_ids[max(0, prompt_len - offset) :].tolist()
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/kimi_audio/kimi_audio_ar_stage.py]: Check scheduler IDs on CPU before transfer; this .tolist() synchronizes every decode step.

File: vllm_omni/model_executor/models/kimi_audio/detokenizer/flow_matching/ode_wrapper.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        self.x_cond = x_cond
+
+        position_ids_cur = [i for i in range(start_position_id, self.x_cond.shape[1] + start_position_id)]
+        position_ids = torch.tensor([position_ids_cur])
+        self.position_ids = position_ids.to(self.x_cond.device).long()
+        self.seq_len = torch.Tensor([position_ids.shape[1]]).to(self.x_cond.device).long()
+
+        cu_seqlens = torch.cumsum(self.seq_len, dim=0)
+        self.cu_seqlens = torch.cat([torch.Tensor([0]).to(cu_seqlens.device), cu_seqlens], dim=0).int()
+        self.cu_maxlen = self.seq_len.cpu().max()
```
Comment (lishunyang12) [inline vllm_omni/model_executor/models/kimi_audio/detokenizer/flow_matching/ode_wrapper.py]: Compute sequence lengths from shapes; these CPU copies synchronize every acoustic chunk.

File: tests/e2e/online_serving/test_kimi_audio.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+]
+
+# Long enough to require more than one acoustic block during ordinary reading.
+TEXT = (
+    "The morning sun shines through the window. I open a book and read a story about a quiet village. "
+    "Outside, children are playing in the garden, and a gentle breeze moves the leaves on the trees."
+)
+
+
+@hardware_test(res={"cuda": "H100"}, num_cards=1)
```
Comment (lishunyang12) [inline tests/e2e/online_serving/test_kimi_audio.py:56]: Add this suite to the ready H100 pipeline with the kimi-audio dependencies.

File: vllm_omni/model_executor/models/kimi_audio/detokenizer/flow_matching/ode_wrapper.py (hunk 3 above)
Comment (HelloWorldU) [reply vllm_omni/model_executor/models/kimi_audio/detokenizer/flow_matching/ode_wrapper.py]: cool

File: vllm_omni/model_executor/models/kimi_audio/kimi_audio_ar_stage.py (hunk 2 above)
Comment (HelloWorldU) [reply vllm_omni/model_executor/models/kimi_audio/kimi_audio_ar_stage.py]: Structural and state checks remain elsewhere in the same method; moving the equality checks to CPU would require extending the shared runner interface. So I dropped them here.

**Comment by HelloWorldU** [comment]
@lishunyang12 Thanks, updated

**Comment by HelloWorldU** [comment]
Ping

**Comment by HelloWorldU** [comment]
Adding a brief author self-review for the standard review process: I reviewed the Kimi-Audio integration and follow-up changes across the model stages, input/output paths, serving/streaming, and test scope. Single-GPU offline/online, streaming, concurrency, and cancellation cases were exercised on an A800. Multi-GPU and performance testing remain unverified. I’ll wait for maintainer feedback.
